docs/design/DAM-on-type-and-RUC-interaction.md

```diff
@@ -0,0 +1,151 @@
+
+## Existing rules
+
+* RUC on method suppresses warnings from the method's body and attributes.
+* RUC on method doesn't suppress warnings about the method usage (call to the method, accessing the method via reflection, ...)
+* RUC on type externally behaves as if all static members and constructors have RUC
+* RUC on type suppresses warnings as if all methods in the type have RUC on them (so method bodies and attributes)
+* RUC on type does NOT suppress warnings about/on nested types in any way
+
+## DAM on type behavior
+
+Annotating type with DAM effectively creates a reflection access to all affected members (depends on the annotation). The behavior in terms of getting a warning or not should be equivalent to accessing the member directly through reflection (`GetMethod` or similar).
+
+For better warning UX and suppression we decided to generate the warnings from:
+
+* The member itself if the type or one of its base types is annotated with a DAM which accesses the member in question
+* The type which adds DAM annotation which accesses a member on a base type (so unlike the above rule, we don't warn on the member, but on the type which added the DAM)
+
+*Note: the reasoning behind this behavior is to warn on the thing which is to blame for the problem. If type derives from annotated type, then the source of the problem is adding a member which will be accessed by such DAM and has issues (like RUC on it). If type adds DAM on itself, but such DAM causes access to a problematic member on base type, then the one to blame is the annotated type.*
+
+## RUC on type and interaction with DAM on type
+
+```csharp
+[DAM(All)]
+class AnnotatedBase {}
+
+[RUC]
+class DerivedWithRUC : AnnotatedBase
+{
+    [RUC]
+    public void MethodWithRUC() {}
+
+    public void MethodWithDAM([DAM(Fields)] Type type) {}
+}
+```
+
+Following the existing rules, accessing `MethodWithRUC` or `MethodWithDAM` via reflection will generate a warning, regardless of the RUC on type.
+Since DAM on type is meant to behave as if accessing the members via reflection, those warnings should not be affected by the RUC on type either.
+
+## RUC on method and interaction with DAM on type
+
+```csharp
+[DAM(All)]
+class AnnotatedBase {}
+
+class Derived : AnnotatedBase
+{
+    [RUC]
+    public void MethodWithRUC() {}
+
+    public void MethodWithDAM([DAM(Fields)] Type type) {}
+
+    [RUC]
+    public void MethodWithRUCAndDAM([DAM(Fields)] Type type) {}
+}
+```
+
+Following the existing rules, accessing `MethodWithRUC` or `MethodWithDAM` via reflection will generate a warning. The same rules also imply that accessing `MethodWithRUCAndDAM` via reflection will also generate a warning (actually two of them).
+RUC on method does not affect usage of the method (other than it should generate a warning on the usage).
+
+## RUC or DAM on nested type
+
+```csharp
+[DAM(All)]
+class AnnotatedBase {}
+
+class Derived : AnnotatedBase
+{
+    [RUC]
+    class NestedWithRUC {}
```

Comment:
This goes to the thing I'm not 100% sure, that is that if the nested type is included in the DAM, it will be included as All. If that is always the case, I agree. I'm wondering if we should warn regardless - to not rely on the potentially fragile behavior of marking nested types as All always.

Could you add one more test case - type which has no members - it should get a warning on the default constructor, which will not be visible in C# so it might be tricky to validate with test infra though. Maybe it's not that important (as in IL the type will always have some .ctor, doesn't matter which)

Reply:
As far as I know that is always the case, and if not I would consider it a bug based on the discussion around https://github.com/mono/linker/pull/2133. I'll add the test you suggested, thanks!

If the nested type marking behavior changes so that it doesn't mean All - particularly if the constructor isn't kept - then I wouldn't necessarily expect this to warn either.